src/BenchmarkDotNet.TestAdapter/VSTestAdapter.cs

```diff
@@ -0,0 +1,206 @@
+        /// <param name="frameworkHandle">An interface used to communicate with the VSTest host.</param>
+        /// <param name="testCases">
+        /// The specific test cases to be run if specified.
+        /// If unspecified, runs all the test cases in the source.
+        /// </param>
+        private void RunBenchmarks(string source, IFrameworkHandle frameworkHandle, IEnumerable<TestCase>? testCases = null)
+        {
+            ValidateSourceIsAssemblyOrThrow(source);
+
+            // Create a HashSet of all the TestCase IDs to be run if specified.
+            var caseIds = testCases == null ? null : new HashSet<Guid>(testCases.Select(c => c.Id));
+
+            try
+            {
+                // Ensure that test execution is done inside the context of the source directory.
+                var executor = (BenchmarkExecutorWrapper)CreateIsolatedType(typeof(BenchmarkExecutorWrapper), source);
+                cts?.Token.Register(executor.Cancel);
+
+                executor.RunBenchmarks(source, new TestExecutionRecorderWrapper(frameworkHandle), caseIds);
+            }
+            catch (Exception ex)
+            {
+                frameworkHandle.SendMessage(TestMessageLevel.Error, $"Failed to run benchmarks in assembly\n{ex}");
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// This will create the given type in a child AppDomain when used in .NET Framework.
+        /// If not in the .NET Framework, it will use the current
+        /// </summary>
+        /// <param name="type">The type to create.</param>
+        /// <param name="assemblyPath">The dll or exe of the benchmark project.</param>
+        /// <returns>The created object.</returns>
+        private static object CreateIsolatedType(Type type, string assemblyPath)
+        {
+#if NETFRAMEWORK
+            var appBase = Path.GetDirectoryName(assemblyPath);
+            var setup = new AppDomainSetup { ApplicationBase = appBase };
+            var domainName = $"Isolated Domain for {type.Name}";
+            var appDomain = AppDomain.CreateDomain(domainName, null, setup);
+            return appDomain.CreateInstanceAndUnwrap(
+                type.Assembly.FullName, type.FullName, false, BindingFlags.Default, null, null, null, null);
+#else
+            return Activator.CreateInstance(type);
+#endif
+        }
+
+        private static void ValidateSourceIsAssemblyOrThrow(string source)
+        {
+            if (string.IsNullOrEmpty(source))
+                throw new ArgumentException($"'{nameof(source)}' cannot be null or whitespace.", nameof(source));
+
+            if (!Path.HasExtension(source))
+                throw new NotSupportedException($"Missing extension on source '{source}', must have the extension '.dll' or '.exe'.");
+
+            var extension = Path.GetExtension(source);
+            if (!string.Equals(extension, ".dll") && !string.Equals(extension, ".exe"))
```

Comment:
It will be safer to ignore casing in this comparison (extensions `.DLL` and `.EXE` are not common, but they do exist).